### PR TITLE
Users step

### PR DESCRIPTION
**Pull request by rvykydal**
This is a draft pull request for https://issues.redhat.com/browse/INSTALLER-3205.
I have some questions (some related to specific commits, so it might make sense to follow the PR with the commits) and items for follow up steps.

- [x] Figure out where to keep the new UI state. In the draft I picked a structure in wizard - see some reasoning and options in the commit message [webui: keep the state of Create Account UI](https://github.com/rhinstaller/anaconda/pull/5280/commits/1f9f776cf58b3afc88526102770a83bed693bec6).
- [x] [REQUIRED IN THIS PR] Strength indicator update / redesign. As can be seen on the screenshot the current indicator (following PF example) does not scale for horizontal forms (compare users and disk encryption sshots). https://github.com/patternfly/patternfly/issues/6032
- [ ] Add user name validation. Currently only non-zero length is required. The rules in Gtk UI are pretty numerous and seem to need review (and maybe compare with gnome tools).
- [ ] [~~REQUIRED IN THIS PR~~] Add account settings to the review screen (perhaps also when root account part is finished).
- [ ] Crypt user password before passing it to backend (as in Gtk UI).
- [x] [REQUIRED IN THIS PR] Add tests, remove hack [webui: FIXME do not require user temporarily](https://github.com/rhinstaller/anaconda/pull/5280/commits/5511082b0a200e9bff71c39c76d37e9eb2135c09). The tests will need nontrivial update for a new mandatory step with required user interaction.
- [x] [REQUIRED IN THIS PR] run only for non-workstation case 
- [ ] make sure the tests work both for workstation / non-workstation (live/non-live) case - this would involve replacing back() back() with something like reach_back(STEP_ID)
- [x] better prefixId setting in the components
- [ ] add thorough user screen test (going back and forth, rule checkers, pwd strength,...) - perhaps when also root accout part is added
- [x] Fix e2e tests
- [ ] Add the hint to the description.

Screenshot from 2023-10-25 15-39-02

The strenght indicator does not scale for isHorizontal Form
Screenshot from 2023-10-25 15-39-19
Screenshot from 2023-10-25 15-39-29
Compare to disk encryption:
Screenshot from 2023-10-25 15-39-40

### CODE REVIEW

File: ui/webui/src/components/Password.jsx
Comment (KKoukiou) [inline ui/webui/src/components/Password.jsx:35]: Please use named imports from @patternfly/react-icons

import { IconONe, IconTwo } from "@patternfly/react-icons"

**Review by KKoukiou — COMMENTED** [review]
Will review this tomorrow morning.

File: ui/webui/src/components/storage/DiskEncryption.jsx
Comment (KKoukiou) [inline ui/webui/src/components/storage/DiskEncryption.jsx]: satisfied is a boolean like the isWarning variable. For consistency purposes, let's pick one and use either the pattern 'isMyBoolVariable' or 'myBoolVariable'.

File: ui/webui/src/components/storage/DiskEncryption.jsx
Comment (KKoukiou) [inline ui/webui/src/components/storage/DiskEncryption.jsx]: I think it's more explicit to use `isError: false` instead of `isWarning: true`. and vice versa. I had to read the code through to understand that `isWarning: true` is not causing the password to be invalid.

File: ui/webui/src/components/storage/DiskEncryption.jsx
Comment (KKoukiou) [inline ui/webui/src/components/storage/DiskEncryption.jsx:74]: I don't see the purpose of maintaining the isValid as a state variable here. I suggest to directly use the setIsFormValid to invalidate or validate the form from the children component and remove this.

* DiskEncryption should always setIsFormValid(true) if the isEncrypted is false.
* The Password form fields should setIsFormValid(satisfied) where satisfied is the password rules result.

File: ui/webui/src/components/users/Accounts.jsx
Comment (KKoukiou) [inline ui/webui/src/components/users/Accounts.jsx:101]: Instead of using Stack/StackItems you can put everything in a `<Form isHorizontal>`, spacings will be correct automatically.

File: ui/webui/src/components/AnacondaWizard.jsx
Comment (KKoukiou) [inline ui/webui/src/components/AnacondaWizard.jsx:363]: Hm does this need special condition? The else { does the same.

**Review by KKoukiou — CHANGES_REQUESTED** [review]

File: ui/webui/src/components/users/Accounts.jsx
Comment (rvykydal) [reply ui/webui/src/components/users/Accounts.jsx:101]: I am using the Stack to get the spacing between rows (like in the mockup), without it it looks like this:
Screenshot from 2023-10-30 17-18-18

File: ui/webui/src/components/AnacondaWizard.jsx
Comment (rvykydal) [reply ui/webui/src/components/AnacondaWizard.jsx:365]: I guess you overlooked setUsers ?

File: ui/webui/src/components/users/Accounts.jsx
Comment (KKoukiou) [reply ui/webui/src/components/users/Accounts.jsx:101]: The proper way to space form elements it to wrap everything with a '<Form isHorizontal>' and let Patternfly do the trick :)

Comment (rvykydal) [reply ui/webui/src/components/users/Accounts.jsx:101]: Ah, this works if I move Form from `Accounts` to `Create Account` (having it in Accounts does not work). I wanted the Form to contain also (future) root component. I guess I'll use two forms then - `Create Account` and the other one for the root component? Well, it can be addressed when I add the UI for root. For now I'll move Form to `Create Accounts`.

File: ui/webui/src/components/Password.jsx
Comment (rvykydal) [reply ui/webui/src/components/Password.jsx]: Updated

File: ui/webui/src/components/storage/DiskEncryption.jsx
Comment (rvykydal) [reply ui/webui/src/components/storage/DiskEncryption.jsx]: Updated

File: ui/webui/src/components/storage/DiskEncryption.jsx
Comment (rvykydal) [reply ui/webui/src/components/storage/DiskEncryption.jsx]: Updated

File: ui/webui/src/components/storage/DiskEncryption.jsx
Comment (rvykydal) [reply ui/webui/src/components/storage/DiskEncryption.jsx]: Updated

**Comment by rvykydal** [comment]
I think I addressed the remarks above and I am going to work on the items from TODO list in the description.

File: ui/webui/test/check-storage
Comment (rvykydal) [inline ui/webui/test/check-storage:426]: This will now work for cases where the accounts screen is disabled (eg Workstation). We may need something like reach() for going back.

File: ui/webui/src/components/AnacondaWizard.jsx
Comment (rvykydal) [inline ui/webui/src/components/AnacondaWizard.jsx:158]: We should hide the screen on WS (!isBootIso)

**Comment by M4rtinK** [comment]
@rvykydal @KKoukiou So thinking about it. I think the user screen needs to be conventionalized not to show up on Fedora Workstation. They have Gnome Initial Setup for user creation, so the extra screen would change the expected and agreed upon flow (and might break Fedora QA testing if an unexpected screens just shows up).

I think the logic should be kinda like this:

- Live images in general: YES
- Fedora Workstation : NO
- boot.iso: YES

I think this logic should either be part of this PR before its merged or alternatively we should not do a new release before the logic id in place via a followup. :)

**Comment by rvykydal** [comment]
> @rvykydal @KKoukiou So thinking about it. I think the user screen needs to be conventionalized not to show up on Fedora Workstation. They have Gnome Initial Setup for user creation, so the extra screen would change the expected and agreed upon flow (and might break Fedora QA testing if an unexpected screens just shows up).
> 
> I think the logic should be kinda like this:
> 
>     * Live images in general: YES
> 
>     * Fedora Workstation : NO
> 
>     * boot.iso: YES
> 
> 
> I think this logic should either be part of this PR before its merged or alternatively we should not do a new release before the logic id in place via a followup. :)

I think in scope of this PR just using isBootIso logic similar as with Language selection should do (would disable the users screen at live images in general, including Fedora WS), but yes we may soon need to use the per-product/variant logic rather then simple isBootIso.

File: ui/webui/src/components/AnacondaWizard.jsx
Comment (rvykydal) [reply ui/webui/src/components/AnacondaWizard.jsx]: Addressed in [webui: hide user screen on live images](https://github.com/rhinstaller/anaconda/pull/5280/commits/251e53528bf15e18821e282ed71674dd25cc3cae)

**Comment by rvykydal** [comment]
@KKoukiou I think you didn't review the last 8 commits - mostly adding the tests.
I believe the PR is mergeable (see the missing items in the description).

File: ui/webui/test/check-basic
Comment (rvykydal) [inline ui/webui/test/check-basic:67]: I think the callback setting should go to the installer as a default for ACCOUNTS step so that we can just call i.reach(i.steps.REVIEW) from the tests.

**Comment by rvykydal** [comment]
Need to do a tests update.

File: ui/webui/test/check-basic
Comment (rvykydal) [reply ui/webui/test/check-basic]: Done in [webui: create required user when reaching a test step by default](https://github.com/rhinstaller/anaconda/pull/5280/commits/3bf083ad86de46a119a1cd9c7a221ca76127fff7)

**Comment by rvykydal** [comment]
@KKoukiou I think you didn't review the last 7 commits - mostly adding the tests.

Regarding the commits with test updates - in case you look at the commits sequentially just a warning that the
[webui: create required user when reaching a test step by default](https://github.com/rhinstaller/anaconda/pull/5280/commits/3bf083ad86de46a119a1cd9c7a221ca76127fff7) is a change of approach to a previous commit.

I believe the PR is mergeable (see the missing items in the description).

File: ui/webui/src/components/users/Accounts.jsx
Comment (KKoukiou) [inline ui/webui/src/components/users/Accounts.jsx:145]: Uneeded wrapper on empty compoent.

File: ui/webui/src/components/users/Accounts.scss
Comment (KKoukiou) [inline ui/webui/src/components/users/Accounts.scss:1]: Wrong comment.

File: ui/webui/test/check-basic
Comment (KKoukiou) [inline ui/webui/test/check-basic]: I would like that the callback would be by default set if not specified to avoid repeating code.

File: ui/webui/test/check-storage
Comment (KKoukiou) [inline ui/webui/test/check-storage:1045]: This should be added in anacondalib and be run by default for all tests.

**Review by KKoukiou — APPROVED** [review]
All in all I have some comments, but they are all not blockers. This is a huge chunk of work and in very good state. I am ok to merge this as is, and do the improvements in followups.

File: ui/webui/src/components/users/Accounts.jsx
Comment (rvykydal) [reply ui/webui/src/components/users/Accounts.jsx:145]: Yes, it is waiting for the root part. Either I'll keep it or remove when I add root.

File: ui/webui/test/check-basic
Comment (rvykydal) [reply ui/webui/test/check-basic]: Yes, it is fixed in the next commit.

**Review by KKoukiou — COMMENTED** [review]
Agreed with @rvykydal to block this on fixing the strength indicator.

File: ui/webui/test/check-storage
Comment (rvykydal) [reply ui/webui/test/check-storage:1045]: Okay, sounds good, only on destructive tests if failed for me.
I think we can address it in a follow up PR.

**Review by KKoukiou — COMMENTED** [review]
@rvykydal I think I have a workaround:

``` diff
[kkoukiou@sequioa webui]$ git diff
diff --git ui/webui/src/components/users/Accounts.scss ui/webui/src/components/users/Accounts.scss
index f53ec05624..bdce550c8f 100644
--- ui/webui/src/components/users/Accounts.scss
+++ ui/webui/src/components/users/Accounts.scss
@@ -2,3 +2,8 @@
 #accounts-create-account {
   width: min(60ch, 100%);
 }
+
+// FIXME: Undo when fixed upstream: https://github.com/patternfly/patternfly/issues/6032
+#accounts-create-account .pf-v5-c-form__group-label.pf-m-info {
+    align-items: baseline
+}
```


And since it affects also the disk selection, actually the workaround needs to be on the CSS file of the Password component.

**Comment by rvykydal** [comment]
> @rvykydal I think I have a workaround:
> 
> ```diff
> [kkoukiou@sequioa webui]$ git diff
> diff --git ui/webui/src/components/users/Accounts.scss ui/webui/src/components/users/Accounts.scss
> index f53ec05624..bdce550c8f 100644
> --- ui/webui/src/components/users/Accounts.scss
> +++ ui/webui/src/components/users/Accounts.scss
> @@ -2,3 +2,8 @@
>  #accounts-create-account {
>    width: min(60ch, 100%);
>  }
> +
> +// FIXME: Undo when fixed upstream: https://github.com/patternfly/patternfly/issues/6032
> +#accounts-create-account .pf-v5-c-form__group-label.pf-m-info {
> +    align-items: baseline
> +}
> ```
This didn't work for me:
Screenshot from 2023-11-06 09-40-30
But adding also `flex-direction: column` did:
Screenshot from 2023-11-06 09-41-23

> 
> And since it affects also the disk selection, actually the workaround needs to be on the CSS file of the Password component. Add some helper class there and replace my used ID.
Not sure what you mean by the disk selection. I knou about ths issue (using the indicator in horizontal form) only at this place, so the ID would be OK.

**Comment by rvykydal** [comment]
> But adding also `flex-direction: column`
> 
> ```
> +
> +// FIXME: Undo when fixed upstream: https://github.com/patternfly/patternfly/issues/6032
> +#accounts-create-account .pf-v5-c-form__group-label.pf-m-info {
> +  flex-direction: column;
> +  align-items: baseline;
> +}
> ```

Actually on Chrome the `baseline` does not look good, using `flex-start` instead.

**Comment by KKoukiou** [comment]
> > But adding also `flex-direction: column`
> > ```
> > +
> > +// FIXME: Undo when fixed upstream: https://github.com/patternfly/patternfly/issues/6032
> > +#accounts-create-account .pf-v5-c-form__group-label.pf-m-info {
> > +  flex-direction: column;
> > +  align-items: baseline;
> > +}
> > ```
> 
> Actually on Chrome the `baseline` does not look good, using `flex-start` instead.

I don't like the change that the 'direction: column' does. The design seems of when the strength goes bellow imo. Fine with the flex-start.

File: ui/webui/src/components/users/Accounts.scss
Comment (rvykydal) [reply ui/webui/src/components/users/Accounts.scss]: [webui: fix password strength indicator layout in horizontal form](https://github.com/rhinstaller/anaconda/pull/5280/commits/1588dad011eebe7cfdf4d65d6d4a6a697914bd23)

**Review by KKoukiou — APPROVED** [review]
Thanks. As said I dont like the column direction for the strength indicator but I am happy to send a followup fix. Also I think we need to show the new user name in the review step as well? Also fine for followup.

**Comment by rvykydal** [comment]
> Thanks. As said I dont like the column direction for the strength indicator but I am happy to send a followup fix. Also I think we need to show the new user name in the review step as well? Also fine for followup.

Agreed, I have the review step in the TODO list at the beginning. I would merge this. (I've got the feeling we may end up with completely different indicator, similar to the bar in cockpit or Anaconda.)

**Comment by rvykydal** [comment]
Follow up issues tracked here: https://issues.redhat.com/browse/INSTALLER-3781

**Pull request merged**
